dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/InactivityValidationAwareConnectionKeepAliveStrategy.java

```diff
@@ -0,0 +1,88 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.hc5;
+
+import com.palantir.logsafe.SafeArg;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.hc.client5.http.ConnectionKeepAliveStrategy;
+import org.apache.hc.client5.http.impl.DefaultConnectionKeepAliveStrategy;
+import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HeaderElements;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.hc.core5.util.TimeValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class InactivityValidationAwareConnectionKeepAliveStrategy implements ConnectionKeepAliveStrategy {
+    private static final Logger log =
+            LoggerFactory.getLogger(InactivityValidationAwareConnectionKeepAliveStrategy.class);
+
+    private static final ConnectionKeepAliveStrategy DELEGATE = DefaultConnectionKeepAliveStrategy.INSTANCE;
+    private final PoolingHttpClientConnectionManager connectionManager;
+    private final String clientName;
+    private final TimeValue defaultValidateAfterInactivity;
+    /**
+     * This field is used for observability. It's possible, though unlikely, that the value can get out of sync
+     * with the connection manager in some scenarios.
+     */
+    private final AtomicReference<TimeValue> currentValidationInterval;
+
+    InactivityValidationAwareConnectionKeepAliveStrategy(
+            PoolingHttpClientConnectionManager connectionManager, String clientName) {
+        this.connectionManager = connectionManager;
+        this.clientName = clientName;
+        // Store the initial inactivity interval to restore if responses re received without
+        // keep-alive headers.
+        this.defaultValidateAfterInactivity = connectionManager.getValidateAfterInactivity();
+        this.currentValidationInterval = new AtomicReference<>(defaultValidateAfterInactivity);
+    }
+
+    @Override
+    public TimeValue getKeepAliveDuration(HttpResponse response, HttpContext context) {
+        TimeValue result = DELEGATE.getKeepAliveDuration(response, context);
+        if (result != null
+                // Only use keep-alive values from 2xx responses
+                && response.getCode() / 100 == 2) {
+            TimeValue newInterval =
+                    containsKeepAliveHeaderWithTimeout(response) ? result : defaultValidateAfterInactivity;
+            TimeValue previousInterval = currentValidationInterval.getAndSet(newInterval);
+            if (!Objects.equals(previousInterval, newInterval)) {
+                log.info(
```

Comment:
might want to push this down to debug otherwise it could get noisy

Reply:
I added the `currentValidationInterval` getAndSet to avoid noise when the values are equal. I don't expect this to trigger very frequently. Rolling upgrades across versions can cause a little churn, but I don't expect it to be terrible.